network/src/services/dns_seeding/seed_record.rs

```diff
@@ -0,0 +1,273 @@
+use std::net::IpAddr;
+use std::net::SocketAddr;
+use std::str::FromStr;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use hash::sha3_256;
+use lazy_static::lazy_static;
+use p2p::{
+    multiaddr::{multihash::Multihash, Multiaddr, Protocol, ToMultiaddr},
+    secio::PeerId,
+    utils::is_reachable,
+};
+use secp256k1::{key::PublicKey, Message, RecoverableSignature, RecoveryId};
+
+lazy_static! {
+    static ref SECP256K1: secp256k1::Secp256k1<secp256k1::All> = secp256k1::Secp256k1::new();
+}
+const SEP: char = ';';
+
+// Format:
+// ======
+// {ip};{port};{peer_id(base58)};{valid_until(base10)};{signature(base58)}
+
+// Length calculation:
+// ==================
+// ip          : max   39 bytes (2001:0dc5:72a3:0000:0000:802e:3370:73E4)
+// port        : max   5 bytes (65535)
+// peer_id     : max   (32 + 3) * 2 * 0.8 = 56 bytes (base58)
+// valid_until : max   11 bytes (31536000000, 1000 year)
+// signature   : max   65 * 2 * 0.8 = 104 bytes (base58)
+// sep         : exact 4 bytes
+// total       : max   39 + 5 + 56 + 11 + 104 + 4 = 224 bytes
+// txt limit   : 255 bytes (enough)
+
+// Typical txt record:
+// ==================
+//   47.103.65.40;49582;QmbU82jmDbu8AsUfa6bDKPHxTpwnPfcRQrzNPacKcSyM1Y;1574942409;K1vAkHZZ8to5VmjD4eyv65ENLbNa9Tda4Aytd8DE9iipFQanRpcZtSPyRiiGHThRGJPVRD18KAsGb8kV2s2WBK39R
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SeedRecord {
+    ip: IpAddr,
+    port: u16,
+    peer_id: Option<PeerId>,
+    // Future utc timestamp
+    valid_until: u64,
+    pubkey: PublicKey,
+}
+
+impl SeedRecord {
+    pub fn check(&self) -> Result<(), SeedRecordError> {
+        if !is_reachable(self.ip) {
+            return Err(SeedRecordError::InvalidIp(self.ip));
+        }
+
+        if self.port == 0 {
+            return Err(SeedRecordError::InvalidPort(self.port));
+        }
+
+        if SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs()
+            > self.valid_until
+        {
+            return Err(SeedRecordError::SeedTimeout);
+        }
+        Ok(())
+    }
+
+    pub fn decode(record: &str) -> Result<SeedRecord, SeedRecordError> {
+        if record.split(SEP).count() != 5 {
```

Comment:
I would rather collect the result into a `Vec<&str>` here.

Reply:
Sounds good